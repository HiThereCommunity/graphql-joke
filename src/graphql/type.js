// @flow

import DataLoader from 'dataloader';
import {User} from '../models'

export type ID = string;

export type Context = {
  loaders: {
    todoItem: DataLoader<string, ?Object>,
    user: DataLoader<string, ?Object>
  },
  viewer: User
}
