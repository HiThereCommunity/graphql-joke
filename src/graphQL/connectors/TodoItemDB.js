/**
 * Created by dirk-janrutten on 04/09/16.
 */


export type TodoItemStructure = {
    id: string,
    title: string,
    is_done: boolean,
    creatorId: string //reference to the id of the user
}
var uuid = require('node-uuid');

//This is the place where batching takes place related to the database etc.
//In addition, this class can be configured to create logs of errors in the database and crete graphQL friendly error messages.
export class TodoItemDB {

    todoItems: Array<TodoItemStructure>;

    constructor(){
        this.todoItems = [];

        this.todoItems.push({
            id: "123",
            title: "play tennis",
            is_done: false,
            creatorId: "1234"
        })
    }

    getTodo(id: string, creatorId: string): ?TodoItemStructure{

        let todo =  this.todoItems.find(item => item.id === id && item.creatorId === creatorId);
        if (todo == null) return null;
        else return todo;
    }

    getAllTodo(creatorId: string):Array<TodoItemStructure>{
        return this.todoItems.filter(item => item.creatorId === creatorId);
    }

    deleteTodo(id: string, creatorId: string): ?TodoItemStructure{

        let removeIndex:number = this.todoItems.map((item) => item.id).indexOf(id);

        if (removeIndex === -1) return null;

        let todoItemToDelete = this.todoItems[removeIndex];
        if (todoItemToDelete.creatorId !== creatorId) return null;

        this.todoItems.splice(removeIndex, 1);

        return todoItemToDelete;
    }

    writeTodo(title: string, creatorId: string): TodoItemStructure{

        let todoItem: TodoItemStructure = {
            id: uuid.v4(),
            title: title,
            is_done: false,
            creatorId: creatorId
        };

        this.todoItems.push(todoItem);

        return todoItem;
    }

    updateTodo(id: string, is_done: boolean, creatorId: string): TodoItemStructure {

        let updateIndex:number = this.todoItems.map((item) => item.id)
            .indexOf(id);

        if (updateIndex === -1) return null;

        let todoItemToUpdate: TodoItemStructure = this.todoItems[updateIndex];
        if (todoItemToUpdate.creatorId !== creatorId) return null;

        todoItemToUpdate.is_done = is_done;

        return todoItemToUpdate;

    }
}