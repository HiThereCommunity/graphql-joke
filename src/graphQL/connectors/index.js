/**
 * Created by dirk-janrutten on 04/09/16.
 */


export {TodoItemDB, TodoItemStructure} from "./TodoItemDB";