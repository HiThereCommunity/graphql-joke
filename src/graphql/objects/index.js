// @flow

export { default as GraphQLTodoItem } from "./todoItem";
export { default as GraphQLUser } from "./user";
export { nodeField, nodesField } from "./relayNode";
