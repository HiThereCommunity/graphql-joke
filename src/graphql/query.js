// @flow
import {
  GraphQLObjectType,
  GraphQLNonNull
} from 'graphql'

import {
  nodeField,
  GraphQLTodoList
} from './objects'

import {TodoList} from '../models'
import type {RootValue} from '../types'

export default new GraphQLObjectType({
  name: 'Query',
  description: 'The query root.',
  fields: () => ({
    node: nodeField,
    todoList: {
      type: new GraphQLNonNull(GraphQLTodoList),
      description: 'The list of todo items',
      resolve: (root: RootValue): TodoList => new TodoList(root)
    }
  })
})
