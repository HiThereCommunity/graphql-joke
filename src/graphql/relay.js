// @flow

import {
  nodeDefinitions,
  fromGlobalId
} from 'graphql-relay';

import {
  TodoItem,
  TodoList
} from '../models';

/**
 * We get the node interface and field from the relay library.
 *
 * The first method is the way we resolve an ID to its object. The second is the
 * way we resolve an object that implements node to its type.
 */
export const { nodeInterface, nodeField } = nodeDefinitions(
  (globalId, context, {rootValue}) => {
    const { type, id } = fromGlobalId(globalId);

    if (type === 'TodoItem') {
      return TodoItem.gen(id, rootValue);
    }
    if (type === 'TodoList') {
      return TodoList.gen(id, rootValue);
    }
    return null;
  }
);
