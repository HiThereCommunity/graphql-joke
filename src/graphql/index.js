// @flow

import { GraphQLSchema } from "graphql";

import query from "./query";
import mutation from "./mutations";

export const schema = new GraphQLSchema({
  query,
  mutation
});

export type { Context } from "./type";
