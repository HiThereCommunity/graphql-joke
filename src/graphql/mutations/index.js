// @flow

import {GraphQLObjectType} from 'graphql'
import GraphQLMutationCreateTodo from './addTodo'
import GraphQLMutationUpdateTodo from './updateTodo'
import GraphQLMutationDeleteTodo from './deleteTodo'

export default new GraphQLObjectType({
  name: 'Mutation',
  fields: () => ({
    addTodo: GraphQLMutationCreateTodo,
    updateTodo: GraphQLMutationUpdateTodo,
    deleteTodo: GraphQLMutationDeleteTodo
  })
})
