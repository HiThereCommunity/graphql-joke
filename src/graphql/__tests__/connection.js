
import { testDBConnection } from '../../testUtils'
import {PostgresConnector} from '../../connectors'
import { graphql } from 'graphql'

import schema from '../'

describe('GraphQL connections', () => {
  const postgres = new PostgresConnector(testDBConnection, true)

  const rootValue = {
    db: postgres
  }

  beforeAll(async () => {
    // clear the database before each test. Make sure that the name of the database table matches test.
    await postgres.getSequelize().sync({force: true, match: /test$/})

    const data = [
      { id: 1, title: 'todo 1', completed: false },
      { id: 2, title: 'todo 2', completed: true },
      { id: 3, title: 'todo 3', completed: false },
      { id: 4, title: 'todo 4', completed: false }
    ]

    await postgres.getTodoItemEntity().bulkCreate(data, { validate: true })
  })

  afterAll(() => postgres.getSequelize().close())

  it('fetches all the todos in the list', async () => {
    const query = `
      query TodoListQuery {
        todoList {
          items {
            ...TodoItemConnectionQuery
          }
          completed {
            ...TodoItemConnectionQuery
          }
          active {
            ...TodoItemConnectionQuery
          }
        }
      }

      fragment TodoItemConnectionQuery on TodoItemConnection {
        edges {
          cursor
          node {
            id
            title
            completed
          }
        }
      }
    `

    const result = await graphql(schema, query, rootValue)
    expect(result).toMatchSnapshot()
  })

  it('fetches the first todo in the list after a cursor', async () => {
    const query = `
      query FirstAfterCursorTodoQuery {
        todoList {
          items(first: 1 after: "YXJyYXljb25uZWN0aW9uOjA=") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              node {
                id
                title
                completed
              }
            }
          }
        }
      }
    `

    const result = await graphql(schema, query, rootValue)
    expect(result).toMatchSnapshot()
  })

  it('fetches no todos at the end of the list', async () => {
    const query = `
      query EndOfTodoQuery {
        todoList {
          items(first: 1 after: "YXJyYXljb25uZWN0aW9uOjM=") {
            edges {
              node {
                id
                title
                completed
              }
            }
          }
        }
      }
    `

    const result = await graphql(schema, query, rootValue)
    expect(result).toMatchSnapshot()
  })
})
