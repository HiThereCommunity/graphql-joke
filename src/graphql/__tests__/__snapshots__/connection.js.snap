exports[`GraphQL connections fetches all the todos in the list 1`] = `
Object {
  "data": Object {
    "todoList": Object {
      "active": Object {
        "edges": Array [
          Object {
            "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
            "node": Object {
              "completed": false,
              "id": "VG9kb0l0ZW06MQ==",
              "title": "todo 1",
            },
          },
          Object {
            "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
            "node": Object {
              "completed": false,
              "id": "VG9kb0l0ZW06Mw==",
              "title": "todo 3",
            },
          },
          Object {
            "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
            "node": Object {
              "completed": false,
              "id": "VG9kb0l0ZW06NA==",
              "title": "todo 4",
            },
          },
        ],
      },
      "completed": Object {
        "edges": Array [
          Object {
            "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
            "node": Object {
              "completed": true,
              "id": "VG9kb0l0ZW06Mg==",
              "title": "todo 2",
            },
          },
        ],
      },
      "items": Object {
        "edges": Array [
          Object {
            "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
            "node": Object {
              "completed": false,
              "id": "VG9kb0l0ZW06MQ==",
              "title": "todo 1",
            },
          },
          Object {
            "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
            "node": Object {
              "completed": true,
              "id": "VG9kb0l0ZW06Mg==",
              "title": "todo 2",
            },
          },
          Object {
            "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
            "node": Object {
              "completed": false,
              "id": "VG9kb0l0ZW06Mw==",
              "title": "todo 3",
            },
          },
          Object {
            "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
            "node": Object {
              "completed": false,
              "id": "VG9kb0l0ZW06NA==",
              "title": "todo 4",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`GraphQL connections fetches no todos at the end of the list 1`] = `
Object {
  "data": Object {
    "todoList": Object {
      "items": Object {
        "edges": Array [],
      },
    },
  },
}
`;

exports[`GraphQL connections fetches the first todo in the list after a cursor 1`] = `
Object {
  "data": Object {
    "todoList": Object {
      "items": Object {
        "edges": Array [
          Object {
            "node": Object {
              "completed": true,
              "id": "VG9kb0l0ZW06Mg==",
              "title": "todo 2",
            },
          },
        ],
        "pageInfo": Object {
          "endCursor": "YXJyYXljb25uZWN0aW9uOjE=",
          "hasNextPage": true,
          "hasPreviousPage": false,
          "startCursor": "YXJyYXljb25uZWN0aW9uOjE=",
        },
      },
    },
  },
}
`;
